### PR TITLE
encode disk image purpose in extension-release.d + os-release

### PR DESCRIPTION
**Pull request by poettering**
Let's make sure our disk images are recognizable so that we can safely distinguish:

1. system OS images (i.e. those you can boot on baremetal/nspawn)
2. initrd images
3. portable service images
4. system extension images for any of the three above

This adds two things:

1. A field SYSEXT_SCOPE for extension-release.d/. If specified can be used to control which scope a system extension is for, i.e. for system OS images, for initrd images or for portable service images.
2. A field PORTABLE_PREFIXES= for os-release. If specified it can be used to declare which suitable portable service prefixes are supported by the image.

Both fields are optional.

With this in place properly built images can now be recognized as initrd, as portable service image and as system extension (for any of the three kinds). Images suitable as system OS images can be detect less reliably: if an init system is included one may assume it's bootable (of course an init system might be in the image purely as effect of poor packaging, i.e. pulled in by dep, but I guess this can be understood as bug in the packaging of the OS).

The `systemd-dissect` tool is updated to make use of this information.

Note that while this adds metainfo for detecting what purposes an image is for this does not restrict the ability to combine multiple of these purposes. For example, an image can be tagged as multiple of the types above, without restrictions.

### CODE REVIEW

**Comment by poettering** [comment]
@keszybz note that with this PR applied sysext images for the initrd must be marked via SYSEXT_SCOPE=initrd, otherwise we'll by default ignore them in systemd-sysext (as the implicit default for SYSEXT_SCOPE= is "system portable", i.e. it does not include "initrd").

File: src/shared/dissect-image.c [hunk 1]
```diff
@@ -3022,6 +3028,7 @@ int dissected_image_acquire_metadata(DissectedImage *m) {
                 [META_OS_RELEASE]        = ("/etc/os-release\0"
                                            "/usr/lib/os-release\0"),
                 [META_EXTENSION_RELEASE] = "extension-release\0", /* Used only for logging. */
+                [META_HAS_INIT_SYSTEM]   = "has-init-system\0",
```
Comment (bluca) [inline src/shared/dissect-image.c]: same comment as above?

File: src/shared/dissect-image.c [hunk 2]
```diff
@@ -3105,28 +3116,60 @@ int dissected_image_acquire_metadata(DissectedImage *m) {
...
+                                break;
+
+                        case META_HAS_INIT_SYSTEM: {
+                                bool found = false;
+                                const char *init;
+
+                                FOREACH_STRING(init,
+                                               "/usr/lib/systemd/systemd",  /* systemd on /usr merged system */
+                                               "/lib/systemd/systemd",      /* systemd on /usr non-merged systems */
+                                               "/sbin/init") {              /* traditional path the Linux kernel invokes */
```
Comment (bluca) [inline src/shared/dissect-image.c:3128]: /usr/sbin/init?

File: src/shared/extension-release.c [hunk 3]
```diff
@@ -25,6 +26,28 @@ int extension_release_validate(
...
+                        if (!extension_sysext_scope_list)
+                                return -ENOMEM;
+                }
+
+                /* by default extension are good for attachment in portable service and on the system */
+                valid = strv_contains(
+                                extension_sysext_scope_list ?: STRV_MAKE("system", "portable"),
+                                host_sysext_scope);
+                if (!valid) {
+                        log_debug("Extension '%s' is not suitable for this scope, ignoring extension.", name);
```
Comment (bluca) [inline src/shared/extension-release.c]: log the scope too?

**Comment by mrc0mmand** [comment]
This makes the CI sad:

```
[   25.116350] systemd-portabled[347]: ../build/src/portable/portable.c:689:9: runtime error: store to null pointer of type 'char **'
[   26.416462] systemd-portabled[347]:     #0 0x563d35715dc3 in portable_extract /systemd-meson-build/../build/src/portable/portable.c:689:29
[   26.432322] systemd-portabled[347]:     #1 0x563d3572c762 in bus_image_common_get_metadata /systemd-meson-build/../build/src/portable/portabled-image-bus.c:159:13
[   26.442761] systemd-portabled[347]:     #2 0x563d3572b350 in redirect_method_to_image /systemd-meson-build/../build/src/portable/portabled-bus.c:216:16
[   26.445942] systemd-portabled[347]:     #3 0x563d35729a58 in method_get_image_metadata /systemd-meson-build/../build/src/portable/portabled-bus.c:224:16
[   27.143951] systemd-portabled[347]:     #4 0x7f8141cd5a91 in method_callbacks_run /systemd-meson-build/../build/src/libsystemd/sd-bus/bus-objects.c:414:21
[   27.144480] systemd-portabled[347]:     #5 0x7f8141cc38eb in object_find_and_run /systemd-meson-build/../build/src/libsystemd/sd-bus/bus-objects.c:1323:21
[   27.144480] systemd-portabled[347]:     #6 0x7f8141cc2925 in bus_process_object /systemd-meson-build/../build/src/libsystemd/sd-bus/bus-objects.c:1443:21
[   27.203983] systemd-portabled[347]:     #7 0x7f8141d283fe in process_message /systemd-meson-build/../build/src/libsystemd/sd-bus/sd-bus.c:2963:13
[   27.204422] systemd-portabled[347]:     #8 0x7f8141d25daa in process_running /systemd-meson-build/../build/src/libsystemd/sd-bus/sd-bus.c:3005:13
[   27.204422] systemd-portabled[347]:     #9 0x7f8141d17810 in bus_process_internal /systemd-meson-build/../build/src/libsystemd/sd-bus/sd-bus.c:3225:21
[   27.206563] systemd-portabled[347]:     #10 0x7f8141d134a8 in sd_bus_process /systemd-meson-build/../build/src/libsystemd/sd-bus/sd-bus.c:3252:16
[   27.208549] systemd-portabled[347]:     #11 0x7f8141d19ce8 in io_callback /systemd-meson-build/../build/src/libsystemd/sd-bus/sd-bus.c:3603:13
[   27.284259] systemd-portabled[347]:     #12 0x7f8141e97d6f in source_dispatch /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:3598:21
[   27.284727] systemd-portabled[347]:     #13 0x7f8141e96400 in sd_event_dispatch /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:4183:21
[   27.284727] systemd-portabled[347]:     #14 0x7f8141e9939c in sd_event_run /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:4244:21
[   27.296183] systemd-portabled[347]:     #15 0x7f8141892cea in bus_event_loop_with_idle /systemd-meson-build/../build/src/shared/bus-util.c:122:21
[   27.301113] systemd-portabled[347]:     #16 0x563d357361db in manager_run /systemd-meson-build/../build/src/portable/portabled.c:119:16
[   27.302865] systemd-portabled[347]:     #17 0x563d3573570b in run /systemd-meson-build/../build/src/portable/portabled.c:161:13
[   27.302865] systemd-portabled[347]:     #18 0x563d357353ea in main /systemd-meson-build/../build/src/portable/portabled.c:168:1
[   27.359762] systemd-portabled[347]:     #19 0x7f8141117b24 in __libc_start_main (/usr/lib/libc.so.6+0x27b24)
[   27.360449] systemd-portabled[347]:     #20 0x563d35714f1d in _start (/usr/lib/systemd/systemd-portabled+0x20f1d)
[   27.360449] systemd-portabled[347]: SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../build/src/portable/portable.c:689:9 in


[   27.422026] systemd-portabled[474]: ==474==ERROR: LeakSanitizer: detected memory leaks
[   27.426033] systemd-portabled[474]: Direct leak of 67584 byte(s) in 1 object(s) allocated from:
[   27.426033] systemd-portabled[474]:     #0 0x7f226162cb19  (/usr/lib/clang/12.0.1/lib/linux/libclang_rt.asan-x86_64.so+0xf4b19)
[   27.426033] systemd-portabled[474]:     #1 0x7f225bdce829  (/usr/bin/../lib/libLLVM-12.so+0xb46829)
[   27.426033] systemd-portabled[474]: SUMMARY: AddressSanitizer: 67584 byte(s) leaked in 1 allocation(s).
```

(not sure about the second one, let's hope it goes away once the UB is resolved)

File: src/portable/portable.c [hunk 4]
```diff
@@ -618,23 +647,27 @@ int portable_extract(
...
+        r = extract_image_and_extensions(
+                        name_or_path,
+                        matches,
+                        extension_image_paths,
+                        /* validate_sysext= */ false,
+                        &image,
+                        &extension_images,
+                        &os_release,
+                        &unit_files,
+                        &valid_prefixes,
```
Comment (bluca) [inline src/portable/portable.c]: validate_sysext= is false here, so this is actually never used

File: src/portable/portable.c [hunk 5]
```diff
@@ -565,16 +569,24 @@ static int extract_image_and_extensions(
                 r = parse_env_file(f, os_release->name,
                                    "ID", &id,
                                    "VERSION_ID", &version_id,
-                                   "SYSEXT_LEVEL", &sysext_level);
+                                   "SYSEXT_LEVEL", &sysext_level,
+                                   "PORTABLE_PREFIXES", &prefixes);
                 if (r < 0)
                         return r;
+
+                if (prefixes) {
```
Comment (bluca) [inline src/portable/portable.c:576]: as the comment above the block says, right now this is used for validation purposes, but this adds to that, so the comment should be updated to reflect that

File: src/dissect/dissect.c [hunk 6]
```diff
@@ -436,6 +441,46 @@ static int action_dissect(DissectedImage *m, LoopDevice *d) {
                     !strv_isempty(m->os_release) ||
                     !strv_isempty(m->extension_release))
                         putc('\n', stdout);
+
+                printf("    Use As: %s%s%s bootable system for UEFI\n",
```
Comment (bluca) [inline src/dissect/dissect.c]: no json-output equivalent?

**Comment by bluca** [comment]
looks very nice, but see comments - also would be good to extend TEST-29-PORTABLE to exercise this functionality

File: src/shared/dissect-image.c (hunk 2 above)
Comment (poettering) [reply src/shared/dissect-image.c:3128]: I am not aware of anything using /usr/sbin/init as path. The kernel default is /sbin/init, and some distros use a kernel config that calls systemd directly (and so does nspawn), but nothing uses /usr/sbin/init afaik. And I wouldn't define anything new here as entrypoint, and reuse what actually is used.

Comment (bluca) [reply src/shared/dissect-image.c:3128]: Yeah was just thinking of getting the real path first on usrmerged systems, but it's not too important

**Comment by poettering** [comment]
Force pushed a new version now. Addresses all issues pointed out and extends the testcase as suggested. PTAL!

File: src/dissect/dissect.c [hunk 7]
```diff
@@ -369,6 +369,17 @@ static int strv_pair_to_json(char **l, JsonVariant **ret) {
         return json_variant_new_array_strv(ret, jl);
 }
 
+static void strv_pair_print(char **l, const char *prefix) {
+        char **p, **q;
+
```
Comment (bluca) [inline src/dissect/dissect.c:375]: strlen will segfault if `prefix` is null, and in other helpers when there's a prefix it's usually optional, so having an assert here would be good

File: src/portable/portable.c [hunk 8]
```diff
@@ -618,23 +650,29 @@ int portable_extract(
                 char **extension_image_paths,
                 PortableMetadata **ret_os_release,
                 Hashmap **ret_unit_files,
+                char ***ret_valid_prefixes,
                 sd_bus_error *error) {
 
         _cleanup_(portable_metadata_unrefp) PortableMetadata *os_release = NULL;
         _cleanup_ordered_hashmap_free_ OrderedHashmap *extension_images = NULL;
         _cleanup_hashmap_free_ Hashmap *unit_files = NULL;
+        _cleanup_(strv_freep) char **valid_prefixes = NULL;
```
Comment (bluca) [inline src/portable/portable.c:659]: this is not passed to extract_image_and_extensions below?

File: src/portable/portable.c [hunk 9]
```diff
@@ -1225,33 +1281,63 @@ int portable_attach(
...
+                        return -ENOMEM;
+
+                valid_prefixes_joined = strv_join(valid_prefixes, ", ");
+                if (!valid_prefixes_joined)
+                        return -ENOMEM;
+
+                return sd_bus_error_setf(
+                                error,
+                                SD_BUS_ERROR_INVALID_ARGS,
+                                "Selected matches %s are not compatible with portable service image '%s%s%s', refusing. (Acceptable prefix matches are: %s)",
```
Comment (bluca) [inline src/portable/portable.c]: maybe quote the matches `matches '%s' are not` for clarity

File: src/dissect/dissect.c [hunk 10]
```diff
@@ -380,6 +382,32 @@ static void strv_pair_print(char **l, const char *prefix) {
         }
 }
 
+static int get_sysext_scopes(DissectedImage *m, char ***ret_scopes) {
+        _cleanup_strv_free_ char **l = NULL;
+        const char *e;
+
+        assert(m);
+
```
Comment (bluca) [inline src/dissect/dissect.c:392]: `assert(ret_scopes)` (or check before deref)

File: src/dissect/dissect.c [hunk 11]
```diff
@@ -757,6 +819,8 @@ static int run(int argc, char *argv[]) {
         _cleanup_(loop_device_unrefp) LoopDevice *d = NULL;
         int r;
 
+        setlocale(LC_ALL, "");
```
Comment (bluca) [inline src/dissect/dissect.c]: why is this added? a comment above would be helpful

File: test/units/testsuite-29.sh [hunk 12]
```diff
@@ -12,6 +12,11 @@ if [[ -v ASAN_OPTIONS || -v UBSAN_OPTIONS ]]; then
     ARGS+=(--profile=trusted)
 fi
 
+systemd-dissect --no-pager /usr/share/minimal_0.raw | grep -q '✓ portable service'
```
Comment (bluca) [inline test/units/testsuite-29.sh:15]: grepping on the --json output might be more reliable?

File: test/test-functions [hunk 13]
```diff
@@ -619,6 +620,7 @@ install_verity_minimal() {
         mkdir -p "$initdir/usr/lib/extension-release.d" "$initdir/usr/lib/systemd/system" "$initdir/opt"
         grep "^ID=" "$os_release" >"$initdir/usr/lib/extension-release.d/extension-release.app0"
         echo "${version_id}" >>"$initdir/usr/lib/extension-release.d/extension-release.app0"
+        echo "SYSEXT_SCOPE=portable" >>"$initdir/usr/lib/extension-release.d/extension-release.app0"
```
Comment (bluca) [inline test/test-functions]: to ensure we don't regress the default, I'd avoid adding this to this image (but ofc leave it in the other one below)

File: src/dissect/dissect.c (hunk 11 above)
Comment (poettering) [reply src/dissect/dissect.c]: it's actually a left-over: i wanted to force systemd-dissect into ASCII mode instead of unicode output mode. The way to do that is setting `LC_CTYPES=de_DE.ISO-8859-1`. But that's only honoured if `setlocale()` is called. 

But as it turns out that requires locales to actually be installed in the testing env, and I figure we shouldn't do that, as all the images would then grow huge. So I went or unicode-based testing, with the unicode check mark. Which I guess is actually better, since it allows us to test the unicode magic by side effect.

Anyway, will drop this here now. In the long run we should probably use setlocale() in all cmdline tools (i.e. not in daemons), so that turning off unicode output works.

File: test/units/testsuite-29.sh (hunk 12 above)
Comment (poettering) [reply test/units/testsuite-29.sh:15]: i was tempted to add `jq` to the tests, but so far none of the tests require that, so i figured it's not the reason here now.

i think the test is much easier to read with the line-based text grep. And it has this nice effect it tests the unicode stuff as side-effect

Comment (bluca) [reply test/units/testsuite-29.sh:15]: yeah for this stuff just a grep of the fixed string should be enough, we have enough testing of the validity of the json formatting

**Comment by poettering** [comment]
All points addressed. PTAL!

**Review by bluca — APPROVED** [review]
lgtm

**Comment by keszybz** [comment]
Late review: LGTM too.

**Comment by keszybz** [comment]
Arrggh, I'm trying to test this with an initrd sysext image, and it's dying with "Failed to mount dissected image: No medium found". No more info despite the debug level.

**Comment by bluca** [comment]
that error usually means it's not happy with os-release/extension-release

**Comment by keszybz** [comment]
I guess so. How can this tool be so user hostile?

**Comment by poettering** [comment]
Did you see:

https://github.com/systemd/systemd/pull/21448#issuecomment-974184113

There should be a very explicit explanation in the debug log at least.

**Comment by poettering** [comment]
In the portabled errors we are pretty explicit on the failure actually. What does `systemctl-dissect` say about the image?

**Comment by keszybz** [comment]
https://github.com/systemd/systemd/pull/21502
